### DISCUSSION TITLE
Which is the correct class type to pass to `Relation` when extending this class?

### DISCUSSION THREAD

**Discussion by nagmat84** [larastan/larastan#1275] [Q&A] [answered]
According to [Upgrade Guide: Generic Relations](https://github.com/nunomaduro/larastan/blob/master/UPGRADE.md#generic-relations) `Relation` is a generic and requires `TRelatedModel`. But what is the correct class type to pass to? A relation relates two models with each other. In particular, I am struggling with many-relations. The return type of the relation should be a collection-like type, but `TRelatedModel` must extend `Illuminate\Database\Eloquent\Model`.

Maybe [Upgrade Guide: Generic Relations](https://github.com/nunomaduro/larastan/blob/master/UPGRADE.md#generic-relations) could be rewritten to be more specific?

**Example**

I have a relation which allows each `Foo` to have up to 6 `Bar`s of different type, i.e. I have a (1:0...6)-relation. The DB looks like this

```sql
CREATE TABLE foos (
  id SERIAL PRIMARY KEY
);

CREATE TABLE bars (
  id SERIAL PRIMARY KEY,
  foo_id INTEGER,
  type SMALLINT NOT NULL,
  UNIQUE KEY id_type_unique (foo_id, type),
  CHECK type (type >= 1 AND type <=6),
  FOREIGN KEY CONSTRAINT foo_id REFERENCES foos(id)
)
```

The unique constraint ensures that for each `foo_id` each type can only be assigned at most once.

I want to implement my own `HasBarBag` relation which does not resolve to a `Collection<Bar>` but to my own `BarBag` which stores up to 6 `B`s acc. to their type. This means I have

```php
class Bar extends Model {
}

class BarBag {
  public ?Bar $type1;
  public ?Bar $type2;
  public ?Bar $type3;
  public ?Bar $type4;
  public ?Bar $type5;
  public ?Bar $type6;
}

/**
 * @property BarBag bar_bag
 */
class Foo extends Model {
  public function bar_bag(): HasBarBag {
    return new HasBarBag($this);
  }
}
```

I wrote the following relation which is works perfectly, but PhpStan is unhappy about it.

```php
/**
 *
 * @phpstan-enxtends HasMany<Bar>     // TODO: this one?
 * @phpstan-enxtends HasMany<Foo>     // TODO: this one?
 * @phpstan-enxtends HasMany<BarBag>  // TODO: this one?
 */
class HasBarBag extends HasMany {
  public function __construct(Foo $foo) {
    parent::__construct(Bar::query(), $foo, 'foo_id', 'id');
  }

  /**
   * Get the results of the relationship.
   *
   * @internal The parent class allows to return `mixed`, hence it is
   * perfectly fine to return `BarBag` acc. to Liskov's substitution
   * principle.
   * However, the  `larastan` ruleset lies to PhpStan about the return
   * type and assumes that `getResult` must always return a model
   * or a collection of models.
   * Hence, we must ignore the false positive here.
   *
   * @return BarBag
   *
   * @phpstan-ignore-next-line
   */
  public function getResults(): BarBag { ... }
}
```

BTW: There is also a bug in `Larastan` with respect to what types might be returned by `getResults`.

**Comment by szepeviktor** [comment]
Hello @nagmat84!
Thank you for starting the discussion.

Up to now I've only seen Models related to Models.
This is our lie: https://github.com/nunomaduro/larastan/blob/master/stubs/HasMany.stub

**Comment by szepeviktor** [comment]
I'm not a developer but you could write: `class BarBag extends Collection`

**Comment by nagmat84** [comment]
Surely not. This would violate Liskov's substitution principle on a completely new level. `BarBag` is not a `Collection` and does _not_ fulfil the "is-a"-relation. For example, you can add arbitrary many items to a `Collection`, but `BarBag` is more a fixed-size array which allows for exactly 6 items. There are incompatible!

**Comment by canvural** [comment]
Hi,

> But what is the correct class type to pass to?

The related model should be passed to relations.

For your example you can use this:
```php
/**
 * @extends HasMany<BarBag>
 */
class HasBarBag extends HasMany {
	// ...
}
```

About the `getResults` type comes from [here.](https://github.com/nunomaduro/larastan/blob/e65978bc2ec0a6177cc1e00d9ed053a877d9e8d3/stubs/HasMany.stub#L14) I think we didn't think about people extending `HasMany` relation, when we add this. We can remove it from stubs probably.

**Reply by nagmat84** [reply]
> The related model should be passed to relations. For your example you can use this:
> 
> ```php
> /**
>  * @extends HasMany<BarBag>
>  */
> class HasBarBag extends HasMany {
> 	// ...
> }
> ```
>

I assume you meant `@extends HasMany<Bar>`. I cannot pass `BarBag` because it does not extend `Model`. The `BarBag` plays the same role as a `Collection` and you do not pass that.

> About the `getResults` type comes from [here.](https://github.com/nunomaduro/larastan/blob/e65978bc2ec0a6177cc1e00d9ed053a877d9e8d3/stubs/HasMany.stub#L14) I think we didn't think about people extending `HasMany` relation, when we add this. We can remove it from stubs probably.

Please note, that the problem goes deeper. It also affects `HasManyOrOne` and `Relation` actually the whole inheritance graph. I guess the correct approach would be to replace `Collection<TRelatedModel>` by `Traversable<TRelatedModel>`.

**Reply by szepeviktor** [reply]
> to replace Collection<TRelatedModel> by Traversable<TRelatedModel>

I like it! Actually how is `BarBag` traversable?

**Reply by nagmat84** [reply]
> > to replace Collection by Traversable
> 
> I like it! Actually how is `BarBag` traversable?

It is not in this example, but there would be no problem to implement the PHP interface. A traversable is not required to support an arbitrary number of elements as a `Collection` does nor is a traversable required to support adding elements to it. A traversable can also be fixed in size. So it would be completely perfect to traverse from element 1 to 6. Hence, `BarBag` can implement `Traversable` without violating Liskov's substitution principle, but it cannot extend `Collection` (because there is no way to add a 7th element).

I assume that it is reasonable to assume that any query result, i.e. any relation set, can be "traversed". Even if its size is limited.

**Comment by szepeviktor** [comment]
Yes, she is watching!

![kép](https://user-images.githubusercontent.com/952007/173247591-bb5598ab-1ea8-4d05-b45f-7f90646a0e10.png)

**Reply by nagmat84** [reply]
I don't get the point of the joke.

**Reply by szepeviktor** [reply]
@nagmat84 She is Liskov!!

**Comment by nagmat84** [comment]
By the way, I found another bug here:

https://github.com/nunomaduro/larastan/blob/ccdd599c464f9de6983a8d5e0f36daaf324b646f/stubs/HasOneOrMany.stub#L11-L20

The parent model (or owning model) is not the same as the related model. Actually, you need to template parameters: `TParentModel` and `TRelatedModel` like this

```php
    /**
     * Create a new has one or many relationship instance.
     *
     * @param  \Illuminate\Database\Eloquent\Builder<TRelatedModel>  $query
     * @param  TParentModel  $parent
     * @param  string  $foreignKey
     * @param  string  $localKey
     * @return void
     */
    public function __construct(\Illuminate\Database\Eloquent\Builder $query, \Illuminate\Database\Eloquent\Model $parent, $foreignKey, $localKey);
```

**Comment by nagmat84** [comment]
I started to work on a PR which fixes the bug above and some related bugs. See there: https://github.com/nagmat84/larastan/tree/fix-relation-types.  Basically, every relation class (`Relation`, `HasOne`, `HasMany`, etc.) needs to distinguish three template parameters `TRelatedModel`, `TOwningModel` and `TResult` which must be kept separately.

I believe I am 90% there. However, some unit tests are still failing. Even though I already have already fixed [`RelationForwardsCallsExtension`](https://github.com/nagmat84/larastan/pull/1/files#diff-edc0a88ec1624e844679e482aaaf80aacb155f4bc6c33c1efc1ae24903684445) and [`ModelRelationsDynamicMethodReturnTypeExtension`](https://github.com/nagmat84/larastan/pull/1/files#diff-af11dc0a6cb05bc6063795b995b97210c60aeee56368b84f649bbb7052210d7b) in order to handle the additional template parameters properly, I still get the following test error

```
There was 1 failure:

1) Tests\FeaturesTest::testFeatures with data set "Models/Relations.php" ('/home/matthias/Entwicklung/la...ns.php')
{
    "totals": {
        "errors": 0,
        "file_errors": 2
    },
    "files": {
        "\/home\/matthias\/Entwicklung\/larastan\/tests\/Features\/Models\/Relations.php": {
            "errors": 2,
            "messages": [
                {
                    "message": "Method Tests\\Features\\Models\\Tag::addresses() should return Illuminate\\Database\\Eloquent\\Relations\\MorphToMany<Tests\\Features\\Models\\Address, Tests\\Features\\Models\\Tag> but returns Illuminate\\Database\\Eloquent\\Relations\\MorphToMany<Tests\\Features\\Models\\Address>.",
                    "line": 322,
                    "ignorable": true
                },
                {
                    "message": "Method Tests\\Features\\Models\\Tag::addressesWithPivot() should return Illuminate\\Database\\Eloquent\\Relations\\MorphToMany<Tests\\Features\\Models\\Address, Tests\\Features\\Models\\Tag> but returns Illuminate\\Database\\Eloquent\\Relations\\MorphToMany<Tests\\Features\\Models\\Address>.",
                    "line": 330,
                    "ignorable": true
                }
            ]
        }
    },
    "errors": []
}
```

Or a little bit more concise:

```
should return MorphToMany<Models\Address, Models\Tag>
but returns MorphToMany<Models\Address>
```

So there must still some code which is not prepared to handle more than one template parameter. Before I had fixed `RelationForwardsCallsExtension` and `ModelRelationsDynamicMethodReturnTypeExtension`, there were much more of these errors.

But now I am at a dead end, because I don't find the code which still needs to be adopted. Is there anybody who would like to give me a helping hand? Any support would be highly appreciated!

**Reply by nagmat84** [reply]
I solved this problem myself.

**Comment by nagmat84** [comment]
@szepeviktor Thank you for leaving a comment on my "internal" PR for https://github.com/nagmat84/larastan/tree/fix-relation-types. While I appreciate your feedback please keep in mind that this is not the real PR. I only created that one for my own tests. The real PR will eventually be against https://github.com/nunomaduro/larastan/tree/1.x.

As I said, I could need a helping hand to sort out the last remaining problems. What shall I do? Create the final PR, mark it as a draft and ask for help there?

**Reply by szepeviktor** [reply]
I think PR-s for v1 won't be accepted.

**Reply by nagmat84** [reply]
But v2 is for Laravel 9 only, right? Our project still requires Laravel 8 due to some 3rd party dependencies which we use and which have not yet migrated to Laravel 9.

**Reply by szepeviktor** [reply]
Yes, now we are working on Larastan v2.

**Reply by szepeviktor** [reply]
> some 3rd party dependencies

Try Laravel Shift service to migrate your dependency to L9.

**Reply by nagmat84** [reply]
Migration of our dependencies to L9 isn't that easy.

So I should write my PR for L9, see if I can get some help to fix the remaining issues and then backport it myself?

What would be the right approach to get help for a PR for L9? Submit what I have, mark it as draft and hope somebody picks it up?

**Reply by szepeviktor** [reply]
> Submit what I have, mark it as draft and hope somebody picks it up?

Yes. And please wait for @canvural. I'm just a janitor here.

**Reply by nagmat84** [reply]
I actually was able to solve the problem myself which I had ran into earlier. Moreover, I have already been able to port everything to the master branch for v2. All tests work, `composer test` is happy and I am currently writing up the text for the PR. There is no need for a draft.

**Comment by nagmat84** [comment, marked as answer]
The ultimate answer to the original question of this post lies in https://github.com/nunomaduro/larastan/pull/1285.

**Reply by szepeviktor** [reply]
I like ultimate solutions.
